env/generate_uuid.cc

```diff
@@ -0,0 +1,129 @@
+//  Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "env/generate_uuid.h"
+
+#ifdef OS_WIN
+#include <rpc.h>  // for uuid generation
+#endif
+
+#include <array>
+#include <atomic>
+#include <random>
+
+#include "rocksdb/env.h"
+#include "rocksdb/file_system.h"
+#include "util/xxhash.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+RawUuid GenerateRandomRawUuidExpensive(Env* env) {
+  // All the data we want to go into the initial random seed. For safety in
+  // case of API issues, we have three entropy tracks that should each be
+  // sufficient for generating unique ids.
+  using RandType = std::random_device::result_type;
+  static constexpr size_t kNumRandVals =
+      /* generous bits */ 192U / (8U * sizeof(RandType));
+  struct Init {
+    // Track 1: host and thread identifiers and time
+    // TODO? process ID
+    uint64_t thread_id = 0;
+    std::array<char, 64> hostname_buf{};
+    int64_t time = 0;
+    uint64_t nano_time = 0;
+
+    // Track 2: random_device
+    std::array<RandType, kNumRandVals> rand_vals{};
+
+    // Track 3: platform uuid
+    RfcUuid uuid;
+  };
+
+  Init init{};
+
+  // Track 1: host and thread identifiers and time
+  init.thread_id = env->GetThreadID();
+  env->GetHostName(init.hostname_buf.data(), init.hostname_buf.size())
+      .PermitUncheckedError();
+  env->GetCurrentTime(&init.time).PermitUncheckedError();
+  init.nano_time = env->NowNanos();
+
+  // Track 2: random_device
+  std::random_device r;
+  for (auto& val : init.rand_vals) {
+    val = r();
+  }
+
+  // Track 3: platform uuid
+  GenerateRfcUuidFromPlatform(&init.uuid).PermitUncheckedError();
+
+  // Mix it all together (nom nom)
+  XXH128_hash_t state = XXH128(&init, sizeof(init), /*seed*/ 123U);
+
+  return (Unsigned128{state.high64} << 64) | state.low64;
+}
+
+const RawUuid& GetRawUuidForCurrentProcess() {
+  // Lazy initialized into static
+  // BUG/FIXME: does not change on fork()
+  static const RawUuid uuid = GenerateRandomRawUuidExpensive(Env::Default());
+  return uuid;
```

Comment:
Is there a race condition here if two threads attempt this call simultaneously?

The name of the API is kind of off/confusing as well, as none of the entropy is related to the "process" per-se (except perhaps the thread id).